Specify ABI version for shared library

avsej opened the pull request:
Without versioning shared objects it won't be possible to maintain ABI compatiblity.

aardappel commented on CMakeLists.txt:
```diff
@@ -189,7 +189,12 @@ endif()
-  set_target_properties(flatbuffers_shared PROPERTIES OUTPUT_NAME flatbuffers)
+
+  set(FlatBuffers_Library_SONAME_MAJOR "1")
+  set(FlatBuffers_Library_SONAME_FULL "${FlatBuffers_Library_SONAME_MAJOR}.0.0")
```
Given that the current version is 1.7.1, maybe set it to that?
Since I do releases, I'll make a note to update this with all the other versions I have to update :)

avsej replied:
Yes, it could be 1.7.1, I will update PR

avsej commented:
@aardappel, but keep in mind that shared object version has some rules, and you have to update major version on breaking ABI changes, even when the project version is not going to change first number. This is why I started from 1.0.0, because in this case you won't confuse it with package version.

But anyway I have changed it to 1.7.0 as you asked, and dropped small note about it. Feel free to update it if you want.

aardappel commented:
While we obviously try to avoid any API changes, they do happen, as the surface area is pretty big. We don't really track ABI changes at all.

The core of FlatBuffers is headers-only, and was generally made with static linking in mind.. any reason why you need this versioned shared library?

I cannot guarantee that the major version will go up on an ABI change. Is there any point in having the version follow along with FlatBuffers releases then?

avsej commented:
> The core of FlatBuffers is headers-only, and was generally made with static linking in mind.. any reason why you need this versioned shared library?

Static libraries discouraged in Linux distributions, and they will track ABI anyway.

> I cannot guarantee that the major version will go up on an ABI change. Is there any point in having the version follow along with FlatBuffers releases then?

Right, maybe lets just make it 1.0.0.

gwvo commented:
On Mon, Nov 6, 2017 at 9:16 PM, Sergey Avseyev <notifications@github.com>
wrote:

> The core of FlatBuffers is headers-only, and was generally made with
> static linking in mind.. any reason why you need this versioned shared
> library?
>
> Static libraries discouraged in Linux distributions, and they will track
> ABI anyway.
>
This only matters if we wanted to distribute FlatBuffers as a stand-alone
library in a Linux distribution, which I would strongly recommend against.
Again, FlatBuffers is a tiny, mostly header-only, in active development
library. There is no advantage to dynamically link it, only disadvantages.
It does not make sense for it to have the same treatment in Linux as
big/stable things like OpenGL, GTK, FreeType, Zlib etc.

Wouter

avsej commented:
Why do you against distributing this library?

aardappel commented:
@avsej see my explanation in my previous comment.

What distribution are you trying to add this to, and what for?

avsej commented:
This is for Fedora and probably CentOS. Static libraries are discouraged there. Also flatbuffers are not headers-only, so it requires shared object to link.

avsej commented:
And also as far as I can see Gentoo also deploys dynamic library: https://gitweb.gentoo.org/repo/gentoo.git/tree/dev-libs/flatbuffers/flatbuffers-1.7.1.ebuild?id=0c24b5e4bd0c84f038d6fd2b94e65b46afd82947#n28

aardappel commented:
I understand static libraries are discouraged in a Linux distribution. My point is that there's no good reason for them to be available in a Linux distribution at all, since most users should want to link FlatBuffers statically (and get it from github rather than apt-get). That's what everyone so far has been doing.

And yes, FlatBuffers is headers-only for the majority of use cases, no object files required. You only need those for more niche use cases such as doing on-the-fly JSON parsing, or reflection etc.

I didn't know about Gentoo, not sure why they decided to add it. Are there any users of that .so?

aardappel commented:
Either way, I'm not going to stop anyone from doing so, this is just my opinion :)
